### PR TITLE
[WIP] Advent discriminator+losses(adver_loss & gan_loss)

### PR DESCRIPTION
**Pull request by tianyu-z**

### CODE REVIEW

**Comment by vict0rsch** [comment]
@tianyu-z is it ready? if so, you should remove [WIP] from the title

File: omnigan/trainer.py [hunk 1]
```diff
@@ -240,6 +243,7 @@ def setup(self):
         self.output_size = self.latent_shape[0] * 2 ** self.opts.gen.t.spade_n_up
         self.G.set_translation_decoder(self.latent_shape, self.device)
         self.D = get_dis(self.opts, verbose=self.verbose).to(self.device)
+        self.ADVENT_D = get_fc_discriminator().to(self.device)
```
Comment (vict0rsch) [inline omnigan/trainer.py:246]: I liked to have all discriminators in `self.D`, and choosing specific keys to reach different models for different tasks. It made things like saving and sending to device a single straightforward op for all of them at once. I also think it makes the code cleaner if we decide to change advent for some reason.

What do you think @melisandeteng @51N84D @alexrey88 @sashavor ?

I mean that the advent discriminator could be a key in `OmniDiscriminator`. What do you people think?

File: omnigan/trainer.py [hunk 2]
```diff
@@ -306,6 +314,18 @@ def d_opt_step(self):
         else:
             self.d_opt.step()
 
+    def ADVENT_D_opt_step(self):
+        """Run an optimizing step ; if using ExtraAdam, there needs to be an extrapolation
+        step every other step
+        """
+        # TODO!!
+        if "extra" in self.opts.dis.opt.optimizer.lower() and (
```
Comment (vict0rsch) [inline omnigan/trainer.py:322]: what's still to do advent-specific keys?

File: omnigan/trainer.py [hunk 3]
```diff
@@ -372,6 +392,8 @@ def run_epoch(self):
             self.update_g(multi_domain_batch)
             if self.d_opt is not None:
                 self.update_d(multi_domain_batch)
+            if self.ADVENT_D_opt is not None:
```
Comment (vict0rsch) [inline omnigan/trainer.py:395]: if advent_d were a substet of D we wouldn't need that extra step and we'd keep the clean code?

File: omnigan/trainer.py [hunk 4]
```diff
@@ -856,6 +879,70 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
         loss = sum(v for d in disc_loss.values() for k, v in d.items())
         return loss
 
+    def update_ADVENT_D(self, multi_domain_batch, verbose=0):
+        # ? split representational as in update_g
+        # ? repr: domain-adaptation traduction
+        self.ADVENT_D_opt.zero_grad()
+        ADVENT_adver_loss = self.get_ADVENT_adver_loss(multi_domain_batch, verbose)
```
Comment (vict0rsch) [inline omnigan/trainer.py:886]: usually in python only class names start with capital letters, variables should start with lower case letters ;)

File: omnigan/trainer.py [hunk 5]
```diff
@@ -856,6 +879,70 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
...
+        """
+        for batch_domain, batch in multi_domain_batch.items():
+            x = batch["data"]["x"]
+            z = self.G.encode(x)
+            source_label = 0
+            target_label = 1
+            losses = []
+            lossCal = ADVENTSegLoss(self.opts)
+            for i, domain in enumerate(batch_domain):
+                pred = self.ADVENT_D(z.cuda(self.device))
```
Comment (vict0rsch) [inline omnigan/trainer.py:907]: this should be a `to(self.device)`

File: omnigan/trainer.py [hunk 6]
```diff
@@ -856,6 +879,70 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
...
+        """Compute the ADVENT adversarial losses:
+            TODO
+        """
+        for batch_domain, batch in multi_domain_batch.items():
+            x = batch["data"]["x"]
+            z = self.G.encode(x)
+            source_label = 0
+            target_label = 1
+            losses = []
+            lossCal = ADVENTSegLoss(self.opts)
```
Comment (vict0rsch) [inline omnigan/trainer.py:905]: why `Cal` ? what does it stand for?

File: omnigan/trainer.py [hunk 7]
```diff
@@ -856,6 +879,70 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
...
+        """
+        for batch_domain, batch in multi_domain_batch.items():
+            x = batch["data"]["x"]
+            z = self.G.encode(x)
+            source_label = 0
+            target_label = 1
+            losses = []
+            lossCal = ADVENTAdversarialLoss(self.opts, None, self.ADVENT_D)
+            for i, domain in enumerate(batch_domain):
+                pred = self.ADVENT_D(z.cuda(self.device))
```
Comment (vict0rsch) [inline omnigan/trainer.py:935]: `to`

File: omnigan/trainer.py (hunk 1 above)
Comment (tianyu-z) [reply omnigan/trainer.py:246]: Okay, I'll fix that.

File: omnigan/trainer.py (hunk 6 above)
Comment (tianyu-z) [reply omnigan/trainer.py:905]: It means Calculation. Should I use func instead of cal?

Comment (melisandeteng) [reply omnigan/trainer.py:905]: I think it's a bit unclear to have cal, why not just loss or advent_loss ?

Comment (tianyu-z) [reply omnigan/trainer.py:905]: Got it. Thanks Meli

File: omnigan/trainer.py (hunk 1 above)
Comment (51N84D) [reply omnigan/trainer.py:246]: Definitely agree. The code shouldn't be built around ADVENT, but rather ADVENT should just be a loss function / mode that we train the mask decoder with.

File: omnigan/losses.py [hunk 8]
```diff
@@ -254,30 +254,15 @@ def __init__(
     
     def __call__(self, prediction1, prediction2, target):
         loss_calc = BCELoss()
-        if self.opt.train.multi_level == True:
+        if self.opt.dis.ADVENT.multi_level == True:
```
Comment (alexrey88) [inline omnigan/losses.py:257]: Details details but I would maybe call your loss attribute "self.opts" instead of "self.opt". I first thought you were initializing the loss with an optimizer but you're giving it the options dict so I think it would be clearer. :)

**Comment by tianyu-z** [comment]
This is a runable version and I am running the experiement [here](https://www.comet.ml/tianyu-z/omnigan/266d7987bdda4245bfd32d58d1fa1419?experiment-tab=chart&showOutliers=true&smoothing=0&view=Default%20view&xAxis=step). I will start to do the clean-code stuff (combine the ADVENT into the exisiting discriminator dictionary) today.

**Comment by tianyu-z** [comment]
The latest version experiment (a version of ADVENT+DeepLabv2_pretrain):
https://www.comet.ml/tianyu-z/omnigan/0c8fd18c544e4a389d2c7f7bffee210c?experiment-tab=chart&showOutliers=true&smoothing=0&view=Default%20view&xAxis=step

**Comment by tianyu-z** [comment]
[experiment](https://www.comet.ml/tianyu-z/omnigan/16fdde4d77a74c78beafd52477d56db4?experiment-tab=images&groupBy=false&orderBy=desc&sortBy=step)(used deeplabv2 pretrained)

**Comment by tianyu-z** [comment]
new [experiment](https://www.comet.ml/tianyu-z/omnigan/9840717692484e7dbc8fd67abd052031?experiment-tab=images&groupBy=false&orderBy=desc&sortBy=step), hope it won't be shutdown
